scripts/prices.js

```diff
@@ -1,88 +1,136 @@
 import { getEventEmitter } from './event_bus.js';
-import { isEmpty } from './misc.js';
-import { cMarket, fillCurrencySelect } from './settings.js';
+import { cOracle, fillCurrencySelect, strCurrency } from './settings.js';
 
 /**
- * CoinGecko's endpoint for PIVX data, optimised for least bandwidth
- * - No localisation, tickers, community data, developer data or sparklines
+ * @typedef {Object} Currency
+ * @property {string} currency - The type of currency
+ * @property {number} value - The value of the currency
+ * @property {number} last_updated - The timestamp when this value was last updated
  */
-export const COINGECKO_ENDPOINT =
-    'https://api.coingecko.com/api/v3/coins/pivx?localization=false&tickers=false&market_data=true&community_data=false&developer_data=false&sparkline=false';
 
 /**
- * The generic market data source template, used to build site-specific classes
+ * Oracle's primary instance.
+ *
+ * @todo Allow an array of Oracle instances for better privacy and decentralisation
  */
-export class MarketSource {
-    /** The storage object for raw market data */
-    cData = {};
+export const ORACLE_BASE = 'https://pivxla.bz/oracle/api/v1';
 
-    /** The name of the market source */
-    strName = '';
-
-    /** The customised API endpoint of the market source */
-    strEndpoint = '';
+/**
+ * An Oracle instance
+ */
+export class Oracle {
+    /**
+     * The currencies cache list
+     * @type {Currency[]} Array to store currency objects
+     */
+    arrCurrencies = [];
 
     /**
-     * Ensure a market data cache exists, if not, fetch it and resume
+     * A lock-like flag which waits until at least once successful "full fetch" of currencies has occurred.
+     * This flag massively lowers bandwidth by only fetching the bulk once, falling to per-currency APIs afterwards.
      */
-    async ensureCacheExists() {
-        if (!this.cData || !Object.keys(this.cData).length) await this.fetch();
-    }
+    fLoadedCurrencies = false;
```

Comment:
I don't like this logic.

With this flag we end up with a lot of if where we check if we already loaded currencies.

I suggest a better approach in which we make sure that currencies are always loaded (i.e. the flag is always true) for example by calling getCurrencies in the constructor and making sure that the operation is successful.

Reply:
I agree it adds a bit of unnecessary logic/statements throughout the price code, but I have concerns changing it how you described, i.e: What if a constructor-based `getCurrencies` fails permanently due to an Oracle being offline? I'm not sure how MPW would handle this logic safely without a flag or lots of `cache.length !== 0` checks.

In the current system, even if Oracle is dead at construction, MPW will keep trying to fill the cache with `getCurrencies` every 15 seconds, then use the flag to switch to the lower bandwidth `getPrice` endpoint.